### PR TITLE
Map store Authorization_Failure to HTTP Forbidden. (#860)

### PR DESCRIPTION
**Pull request by zzmao**
The mapping chain is:
Store(Authorization_Failure) -> Server(Blob_Authorization_Failure)
-> Router(BlobAuthorizationFailure) -> rest(AccessDenied) -> HTTP(Forbidden).

### CODE REVIEW

File: ambry-router/src/main/java/com.github.ambry.router/GetOperation.java [hunk 1]
```diff
@@ -170,15 +170,16 @@ boolean isOperationComplete() {
 
   /**
    * Set the exception associated with this operation.
-   * A {@link ServerErrorCode#Blob_Deleted} or {@link ServerErrorCode#Blob_Expired} error overrides any other
-   * previously received exception.
+   * A {@link ServerErrorCode#Blob_Deleted} or {@link ServerErrorCode#Blob_Expired} or {@link ServerErrorCode#Blob_Expired}
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/GetOperation.java]: minor, the new one should be `Blob_Authorization_Failure`?

Comment (zzmao) [reply ambry-router/src/main/java/com.github.ambry.router/GetOperation.java]: fixed.

**Comment by vgkholla** [comment]
@zzmao Once you are done with the compaction changes, I think we can resume and update this PR and get it merged. @jsjtzyy also has plans to add a new `ServerErrorCode` (to both frontend and storage server) and we should see if they can be added at once

**Comment by zzmao** [comment]
Code updated. @jsjtzyy is invited for review. 
We can merge this first so that @jsjtzyy can do his change based on this one.

File: ambry-commons/src/main/java/com.github.ambry.commons/ServerErrorCode.java [hunk 2]
```diff
@@ -23,6 +23,7 @@
   No_Error,
   IO_Error,
   Blob_Not_Found,
+  Blob_Authorization_Failure,
```
Comment (jsjtzyy) [inline ambry-commons/src/main/java/com.github.ambry.commons/ServerErrorCode.java]: Could you move `Blob_Authorization_Failure` to the end? It's safe to append new error code to enum.

File: ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java [hunk 3]
```diff
@@ -399,24 +399,54 @@ public void testAndAssert(RouterErrorCode expectedError) throws Exception {
...
     doPut();
     Map<ServerErrorCode, RouterErrorCode> serverErrorToRouterError = new HashMap<>();
     serverErrorToRouterError.put(ServerErrorCode.Blob_Deleted, RouterErrorCode.BlobDeleted);
     serverErrorToRouterError.put(ServerErrorCode.Blob_Expired, RouterErrorCode.BlobExpired);
+    serverErrorToRouterError.put(ServerErrorCode.Blob_Authorization_Failure, RouterErrorCode.BlobAuthorizationFailure);
     for (Map.Entry<ServerErrorCode, RouterErrorCode> entry : serverErrorToRouterError.entrySet()) {
       Map<ServerErrorCode, Integer> errorCounts = new HashMap<>();
       errorCounts.put(ServerErrorCode.Blob_Not_Found, replicasCount - 1);
       errorCounts.put(entry.getKey(), 1);
+      System.out.println(errorCounts);
```
Comment (jsjtzyy) [inline ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java]: Do we need to print it out? If not, we can remove this line.

Comment (zzmao) [reply ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java]: No. This was deleted. For some reason, rebase is not taking this change in. 
A new commit was added and also addressed Enum comment. Good point! Thanks!

File: ambry-router/src/test/java/com.github.ambry.router/RouterTestHelpers.java [hunk 4]
```diff
@@ -140,11 +140,22 @@ static void setServerErrorCodes(ServerErrorCode[] serverErrorCodesInOrder, Parti
...
+    for (int i = numOfNodesEachDatacenter; i < serverErrorCodesInOrder.length; i++) {
+      serverErrorCodesInOperationTrackerOrder[i] =
+          serverErrorCodesInOrder[serverErrorCodesInOrder.length + numOfNodesEachDatacenter - 1 - i];
+    }
     int i = 0;
     for (ReplicaId replica : partition.getReplicaIds()) {
       DataNodeId node = replica.getDataNodeId();
       MockServer mockServer = serverLayout.getMockServer(node.getHostname(), node.getPort());
-      mockServer.setServerErrorForAllRequests(serverErrorCodesInOrder[i++]);
+      mockServer.setServerErrorForAllRequests(serverErrorCodesInOperationTrackerOrder[i++]);
```
Comment (jsjtzyy) [inline ambry-router/src/test/java/com.github.ambry.router/RouterTestHelpers.java]: Maybe you can change this loop in the format of `for(int i = 0; i < partition.getReplicaIds().size(); ++i)`. Thus the order is guaranteed in a safer way. (also guarantee the order of mockServer/node)

File: ambry-router/src/test/java/com.github.ambry.router/DeleteManagerTest.java [hunk 5]
```diff
@@ -195,6 +195,42 @@ public void testBlobExpired() throws Exception {
...
+   * Test the case when one server store responds with {@code Blob_Authorization_Failure}, and other servers
+   * respond with {@code Blob_Not_Found}. The delete operation should be able to resolve the
+   * router error code as {@code Blob_Authorization_Failure}. The order of received responses is the same as
+   * defined in {@code serverErrorCodes}.
+   */
+  @Test
+  public void testBlobAuthorizationFailure() throws Exception {
+    ServerErrorCode[] serverErrorCodes = new ServerErrorCode[9];
+    Arrays.fill(serverErrorCodes, ServerErrorCode.Blob_Not_Found);
+    serverErrorCodes[5] = ServerErrorCode.Blob_Authorization_Failure;
```
Comment (jsjtzyy) [inline ambry-router/src/test/java/com.github.ambry.router/DeleteManagerTest.java]: Is there any specific reason we chose index `5` here to set `Blob_Authorization_Failure`?(If not, we might use a random number here to make it more general)

File: ambry-router/src/test/java/com.github.ambry.router/RouterTestHelpers.java [hunk 6]
```diff
@@ -140,11 +140,22 @@ static void setServerErrorCodes(ServerErrorCode[] serverErrorCodesInOrder, Parti
...
+    // serverErrorCodesInOrder.
+    int numOfNodesEachDatacenter = 3;
+    ServerErrorCode[] serverErrorCodesInOperationTrackerOrder = new ServerErrorCode[serverErrorCodesInOrder.length];
+    for (int i = 0; i < numOfNodesEachDatacenter; i++) {
+      serverErrorCodesInOperationTrackerOrder[i] = serverErrorCodesInOrder[numOfNodesEachDatacenter - 1 - i];
+    }
+    for (int i = numOfNodesEachDatacenter; i < serverErrorCodesInOrder.length; i++) {
+      serverErrorCodesInOperationTrackerOrder[i] =
+          serverErrorCodesInOrder[serverErrorCodesInOrder.length + numOfNodesEachDatacenter - 1 - i];
+    }
```
Comment (jsjtzyy) [inline ambry-router/src/test/java/com.github.ambry.router/RouterTestHelpers.java]: I am thinking if we can replace this piece of code with a more intuitive one. How about this?
`    List<ServerErrorCode> errorCodesInAllDCs = Arrays.asList(serverErrorCodesInOrder);
    List<ServerErrorCode> serverErrorInLocalDC = errorCodesInAllDCs.subList(0, 3);
    List<ServerErrorCode> serverErrorInOtherDCs = errorCodesInAllDCs.subList(3, serverErrorCodesInOrder.length);
    Collections.reverse(serverErrorInLocalDC);
    Collections.reverse(serverErrorInOtherDCs);
    List<ServerErrorCode> serverErrorCodesInOperationTrackerOrder = new ArrayList<>(serverErrorInLocalDC);
    serverErrorCodesInOperationTrackerOrder.addAll(serverErrorInOtherDCs);`

File: ambry-router/src/test/java/com.github.ambry.router/RouterTestHelpers.java (hunk 4 above)
Comment (jsjtzyy) [reply ambry-router/src/test/java/com.github.ambry.router/RouterTestHelpers.java]: use `serverErrorCodesInOperationTrackerOrder.get(i++)` if you adopt my suggestion in previous comment.

File: ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java [hunk 7]
```diff
@@ -417,6 +418,34 @@ public void testErrorPrecedenceWithBlobDeletedAndExpiredCase() throws Exception
     }
   }
 
+  /**
+   * Test the case where servers return different {@link ServerErrorCode} or success, and the {@link GetBlobOperation}
+   * is able to resolve and conclude the correct {@link RouterErrorCode}. The get operation should be able
+   * to resolve the router error code as {@code Blob_Authorization_Failure}.
+   */
+  @Test
+  public void testAuthorizationFailureOverrideAll() throws Exception {
```
Comment (jsjtzyy) [inline ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java]: doc `throws Exception` please

File: ambry-router/src/test/java/com.github.ambry.router/GetBlobInfoOperationTest.java [hunk 8]
```diff
@@ -382,6 +384,30 @@ public void testErrorPrecedenceWithBlobDeletedAndExpiredCase() throws Exception
     }
   }
 
+  /**
+   * Test the case where servers return different {@link ServerErrorCode} or success, and the {@link GetBlobInfoOperation}
+   * is able to resolve and conclude the correct {@link RouterErrorCode}. The get blob operation should be able
+   * to resolve the router error code as {@code Blob_Authorization_Failure}.
+   */
+  @Test
+  public void testBlobAuthorizationFailureOverrideAll() throws Exception {
```
Comment (jsjtzyy) [inline ambry-router/src/test/java/com.github.ambry.router/GetBlobInfoOperationTest.java]: doc `throws Exception` please

File: ambry-router/src/test/java/com.github.ambry.router/RouterTestHelpers.java (hunk 4 above)
Comment (zzmao) [reply ambry-router/src/test/java/com.github.ambry.router/RouterTestHelpers.java]: fixed.

File: ambry-router/src/test/java/com.github.ambry.router/DeleteManagerTest.java (hunk 5 above)
Comment (zzmao) [reply ambry-router/src/test/java/com.github.ambry.router/DeleteManagerTest.java]: This is related to `DELETE_PARALLELISM` which is 3 in this test. The index here should < 9 - DELETE_PARALLELISM to make sure we don't hit `Blob_Not_Found`. `testBlobExpired` is also using index 5.

File: ambry-router/src/test/java/com.github.ambry.router/RouterTestHelpers.java (hunk 6 above)
Comment (zzmao) [reply ambry-router/src/test/java/com.github.ambry.router/RouterTestHelpers.java]: good suggestion!

File: ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java (hunk 7 above)
Comment (zzmao) [reply ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java]: fixed.

File: ambry-router/src/test/java/com.github.ambry.router/GetBlobInfoOperationTest.java (hunk 8 above)
Comment (zzmao) [reply ambry-router/src/test/java/com.github.ambry.router/GetBlobInfoOperationTest.java]: fixed.

File: ambry-api/src/test/java/com.github.ambry/store/MockIdFactory.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
+/**
+ * Copyright 2016 LinkedIn Corp. All rights reserved.
```
Comment (jsjtzyy) [inline ambry-api/src/test/java/com.github.ambry/store/MockIdFactory.java]: use 2018 instead please

**Comment by zzmao** [comment]
@vgkholla Let's merge this and give it a shot?

**Comment by vgkholla** [comment]
@zzmao sorry for the delay - will give a look soon

File: ambry-router/src/main/java/com.github.ambry.router/DeleteOperation.java [hunk 10]
```diff
@@ -199,6 +199,9 @@ void handleResponse(ResponseInfo responseInfo, DeleteResponse deleteResponse) {
         } else {
           // The status of operation tracker will be updated within the processServerError method.
           processServerError(replica, deleteResponse.getError(), deleteResponse.getCorrelationId());
+          if (deleteResponse.getError() == ServerErrorCode.Blob_Authorization_Failure) {
```
[the remark this answers is not in the document]
Comment (zzmao) [reply ambry-router/src/main/java/com.github.ambry.router/DeleteOperation.java]: added.

File: ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java [hunk 11]
```diff
@@ -735,6 +735,10 @@ private void writeMessagesToLocalStoreAndAdvanceTokens(List<ExchangeMetadataResp
             replicationMetrics.blobDeletedOnGetCount.inc();
             logger.trace("One of the blobs to GET is deleted: Remote node: {} Thread name: {} Remote replica: {}",
                 remoteNode, threadName, remoteReplicaInfo.getReplicaId());
+          } else if (partitionResponseInfo.getErrorCode() == ServerErrorCode.Blob_Authorization_Failure) {
+            replicationMetrics.blobAuthorizationFailureCount.inc();
+            logger.trace("One of the blobs authorization failed: Remote node: {} Thread name: {} Remote replica: {}",
```
[the remark this answers is not in the document]
Comment (zzmao) [reply ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java]: fixed.

File: ambry-router/src/test/java/com.github.ambry.router/RouterTestHelpers.java [hunk 12]
```diff
@@ -151,11 +151,22 @@ static void setServerErrorCodes(ServerErrorCode[] serverErrorCodesInOrder, Parti
     if (serverErrorCodesInOrder.length > replicas.size()) {
       throw new IllegalArgumentException("More server error codes provided than replicas in partition");
     }
-    int i = 0;
-    for (ReplicaId replica : partition.getReplicaIds()) {
-      DataNodeId node = replica.getDataNodeId();
+    // Do serverErrorCodes assignment based on Operation Tracker logic so that server error can return in the order of
+    // serverErrorCodesInOrder.
+    int numOfNodesEachDatacenter = 3;
```
[the remark this answers is not in the document]
Comment (zzmao) [reply ambry-router/src/test/java/com.github.ambry.router/RouterTestHelpers.java]: As discussed offline, a queue based server error code implementation will be added in the future.

Added test `routerErrorCodeResolutionTest` similar to the one in `ttlUpdateManagerOperation`.

File: ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java [hunk 13]
```diff
@@ -780,6 +780,10 @@ private void writeMessagesToLocalStoreAndAdvanceTokens(List<ExchangeMetadataResp
             replicationMetrics.blobDeletedOnGetCount.inc();
             logger.trace("One of the blobs to GET is deleted: Remote node: {} Thread name: {} Remote replica: {}",
                 remoteNode, threadName, remoteReplicaInfo.getReplicaId());
+          } else if (partitionResponseInfo.getErrorCode() == ServerErrorCode.Blob_Authorization_Failure) {
+            replicationMetrics.blobAuthorizationFailureCount.inc();
+            logger.error("One of the blobs authorization failed: Remote node: {} Thread name: {} Remote replica: {}",
```
Comment (vgkholla) [inline ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java]: minor: can you also print the keys that were part of the get (should be the `missingKeys`)

File: ambry-router/src/test/java/com.github.ambry.router/RouterTestHelpers.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (vgkholla) [reply ambry-router/src/test/java/com.github.ambry.router/RouterTestHelpers.java]: Can you leave a comment here saying this is temporary?

**Pull request merged**